src/Microsoft.AspNet.OData.Shared/Formatter/EdmLibHelpers.cs

```diff
@@ -291,17 +291,20 @@ public static Type GetClrType(IEdmType edmType, IEdmModel edmModel, IWebApiAssem
             }
 
             string typeName = edmSchemaType.FullName();
-            IEnumerable<Type> matchingTypes = GetMatchingTypes(typeName, assembliesResolver);
 
-            if (matchingTypes.Count() > 1)
+            // Calling "ToList()" to avoid multiple enumerates.
+            IList<Type> matchingTypes = GetMatchingTypes(typeName, assembliesResolver).ToList();
+
+            if (matchingTypes.Count > 1)
             {
                 throw Error.Argument("edmTypeReference", SRResources.MultipleMatchingClrTypesForEdmType,
-                    typeName, String.Join(",", matchingTypes.Select(type => type.AssemblyQualifiedName)));
+                    typeName, String.Join(",", matchingTypes.Select(t => t.AssemblyQualifiedName)));
             }
 
-            edmModel.SetAnnotationValue<ClrTypeAnnotation>(edmSchemaType, new ClrTypeAnnotation(matchingTypes.SingleOrDefault()));
+            Type type = matchingTypes.Count == 0 ? null : matchingTypes[0];
```

Comment:
maybe make this name more explicit: "clrType"?

```suggestion
            Type clrType = matchingTypes.Count == 0 ? null : matchingTypes[0];
```

I generally like to avoid naming types "type" because it is confusing with "Type", but in this case it really is a type. Naming it "clrType" helps clarify that we did mean the clrType, not something else we were calling "type" (like the edmType)

Reply:
#Resolved